### PR TITLE
TBOX-203 enforce test coverage

### PR DESCRIPTION
**Pull request by clint-richardson**
# ↪️ Pull Request

## ✔️ PR Todo

- [ ] Add documentation
- [ ] Add examples
- [ ] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [ ] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [ ] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [ ] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [ ] Pass all checks

This PR adds a new job to our `ci.yml` file that checks code test coverage and fails if the total is below 80% and if any individual file has less than 60% coverage

### CODE REVIEW

**Comment by clint-richardson** [comment]
hurray it is failing because our test coverage is below the thresholds I set. I think there are some questions for the maintainers to decide:

* Do we exclude examples from testing?
* Do we exclude the build docs functionality from testing?
* What should the thresholds be for individual files and for total coverage?

let me know and I'll update the PR

**Comment by wpmartin** [comment]
The maintainer team and Sam discussed this and came to a decision. We should add in the exclusions now, omitting the build files and examples from testing. 
Then we should have the coverage thresholds framework in place but do not want to specify anything yet. Looking at the CI output a lot of files had 0% coverage, so we are weary of it being a lot of work to add in for all files. Perhaps the thresholds should be their own ticket? With further tickets to incrementally bump up the thresholds?

**Comment by clint-richardson** [comment]
One solution is to not enforce that single files all are covered, but instead lock in the average (after omitting the examples) so that at least from now on things can only get more tested

**Comment by skalish** [comment]
I'd vote for:

- Average test coverage = <current value> (so it will pass right now)
- Single-file test coverage = 0% (so the framework to increase this is obvious)
- Add `doc_src`, `install.py`, and `setup.py` to the omit list (so it is clear those don't need to be tested, but everything else e.g. examples) does

The check currently seems to be checking if test files themselves are covered, so that may need to be omitted as well.

**Comment by clint-richardson** [comment]
I think the latest commits catch these items (including setting threshold to 0 for single files). I'll check after this runs to see what the test coverage is/what the average threshold should be.

**Comment by skalish** [comment]
It does actually look like min single coverage could be bumped to 10%...

**Comment by skalish** [comment]
@clint-richardson the [documentation](https://github.com/marketplace/actions/pytester-cov) of the plugin you're using is unclear to me. They say the output `cov-threshold-single-fail`: `false if any single file coverage less than cov-threshold-single, else true` which seems to imply it will be True if coverage is okay, but then their example (which I think you followed) interprets `true` as a failure.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -92,3 +92,36 @@ jobs:
...
+      - name: Coverage single fail - exit
+        if: ${{ steps.pytester-cov.outputs.cov-threshold-single-fail == 'true' }}
+        run: |
+          echo "cov single fail ${{ steps.pytester-cov.outputs.cov-threshold-single-fail }}"
+          echo "coverage table ${{ steps.pytester-cov.outputs.output-table }}"
+          exit 1
+      - name: Coverage total fail - exit
+        if: ${{ steps.pytester-cov.outputs.cov-threshold-total-fail == 'true' }}
+        run: |
+          echo "cov single fail ${{ steps.pytester-cov.outputs.cov-threshold-total-fail }}"
```
Comment (skalish) [inline .github/workflows/ci.yml]: Should this be
```suggestion
          echo "cov total fail ${{ steps.pytester-cov.outputs.cov-threshold-total-fail }}"
```

**Review by skalish — CHANGES_REQUESTED** [review]
It looks like there is an echo typo that should be fixed

**Comment by clint-richardson** [comment]
the description in [the codebase](https://github.com/programmingwithalex/pytester-cov/blob/main/action.yml#L37-L40) is consistent with the example:

```
  cov-threshold-single-fail:
    description: 'boolean if any single file coverage less than cov-threshold-single'
  cov-threshold-total-fail:
    description: 'boolean if total coverage less than cov-threshold-total'
```

I believe that the usage is the correct thing, and filed a github issue with the creator about it: https://github.com/programmingwithalex/pytester-cov/issues/2

**Review by skalish — APPROVED** [review]
LGTM.
Just a note on commit types that doesn't require any action: I would consider this entire PR to be `ci` commits (+ `chore` when needed). It doesn't actually add any features or fix any bugs in the library so it shouldn't have any commits labeled as `feat` or `fix`, and under the semantic-release convention would not trigger a release.

**Pull request merged**